### PR TITLE
WIP: More defensive file IO

### PR DESCRIPTION
**Pull request by harshad1**
This PR is meant to address #1482

### CODE REVIEW

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 1]
```diff
@@ -192,21 +193,20 @@ public void onClick(View view) {
...
-                                final String tasksString = TodoTxtTask.tasksToString(keep);
-                                _hlEditor.setText(tasksString);
-                                _hlEditor.setSelection(
-                                        StringUtils.getIndexFromLineOffset(tasksString, selStart),
-                                        StringUtils.getIndexFromLineOffset(tasksString, selEnd)
-                                );
+                            final Document doneDoc = new Document(new File(_document.getFile().getParentFile(), callbackPayload));
+                            final Context context = getContext();
+                            String doneStr = doneDoc.loadContent(context);
+                            if (doneStr != null) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:199]: Don't do anything if the read file was null. i.e. error

File: app/src/main/java/net/gsantner/markor/model/Document.java [hunk 2]
```diff
@@ -214,41 +212,46 @@ public static Document fromArguments(Context context, Bundle arguments) {
         return document;
     }
 
-    public synchronized String loadContent(final Context context) {
+    public String loadContent(final Context context) {
+        return loadContent(this, context);
+    }
+
+    public static synchronized String loadContent(final Document document, final Context context) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/model/Document.java:219]: Wrap static synchronized calls - only write one document at a time.

File: app/src/main/java/net/gsantner/markor/model/Document.java [hunk 3]
```diff
@@ -214,41 +212,46 @@ public static Document fromArguments(Context context, Bundle arguments) {
...
-                Log.e(Document.class.getName(), "loadDocument:  File " + getFile() + " not found.");
-                content = "";
+                Log.e(Document.class.getName(), "loadDocument:  File " + file + " not found.");
+                content = null;
             } catch (JavaPasswordbasedCryption.EncryptionFailedException | IllegalArgumentException e) {
                 Toast.makeText(context, R.string.could_not_decrypt_file_content_wrong_password_or_is_the_file_maybe_not_encrypted, Toast.LENGTH_LONG).show();
-                Log.e(Document.class.getName(), "loadDocument:  decrypt failed for File " + getFile() + ". " + e.getMessage(), e);
-                content = "";
+                Log.e(Document.class.getName(), "loadDocument:  decrypt failed for File " + file + ". " + e.getMessage(), e);
+                content = null;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/model/Document.java:238]: error = return null

File: app/src/main/java/net/gsantner/opoc/util/FileUtils.java [hunk 4]
```diff
@@ -55,7 +55,8 @@ public static String readTextFileFast(final File file) {
         } catch (IOException e) {
             e.printStackTrace();
         }
-        return "";
+
+        return null;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/util/FileUtils.java:59]: error = return null

File: app/src/main/java/net/gsantner/markor/activity/DocumentShareIntoFragment.java [hunk 5]
```diff
@@ -201,22 +202,28 @@ public void setText(String text) {
...
+
+            if (content != null) { // Don't do anything if could not load file
+                content = content.trim();
+                content += (content.isEmpty() ? "" : "\n") + separator + _sharedText;
+                document.saveContent(context, content);
+                document.resetModTime(); // Force load of content in document
+
+                if (showEditor) {
+                    // Open at line
+                    document.setInitialLineNumber(StringUtils.getLineOffsetFromIndex(content, content.length() - 1)[0]);
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/activity/DocumentShareIntoFragment.java:220]: maybe INT_MAX  works, so editor just jumps to the end of the file...without calculation

File: app/src/main/java/net/gsantner/markor/activity/DocumentShareIntoFragment.java [hunk 6]
```diff
@@ -201,22 +202,28 @@ public void setText(String text) {
...
+                    showInDocumentActivity(document);
+                }
 
-            if (file != null) {
-                _appSettings.addRecentDocument(file);
+                if (file != null) {
+                    _appSettings.addRecentDocument(file);
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/activity/DocumentShareIntoFragment.java:225]: should be in the editor class

File: app/src/main/java/net/gsantner/markor/model/Document.java (hunk 3 above)
Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/model/Document.java:238]: maybe

```
catch (encryriptonexception) {
}
catch (exception) {
}
```

to catch all errors at this point. the error-or-not info is contained as null

**Comment by gsantner** [comment]
do you think makes sense to release 2.8.3 with the current master, before merging this? I  heard it's pretty fine on master now. Not that we release another buggy version

**Comment by harshad1** [comment]
> do you think makes sense to release 2.8.3 with the current master, before merging this? I heard it's pretty fine on master now. Not that we release another buggy version

Yes. I think we should.

I would like to test / refine these changes much more. This PR is probably only for 2.9 / 2.9.1.

**Comment by gsantner** [comment]
https://github.com/gsantner/markor/releases/tag/v2.8.3

**Comment by harshad1** [comment]
@gsantner So the whole modtime thing is there to prevent unnecessary reads and writes

The downside of using modtime is that some reads / writes may not happen when we need them to - this requires using resetModTime which is brittle, imo.

What are the downsides of just removing the whole modtime mechanism and allowing the multiple reads and writes (we will still have the hash mechanism to prevent duplicate writes inside `Document`)?

**Comment by gsantner** [comment]
> What are the downsides of just removing the whole modtime mechanism and allowing the multiple reads and writes (we will still have the hash mechanism to prevent duplicate writes inside Document)?

Duplicate read = slow

**Comment by gsantner** [comment]
> What are the downsides of just removing the whole modtime mechanism and allowing the multiple reads and writes

Duplicate read = slow. But if you think it's too much hassle - we can remove it again. Though we need to make sure we don't make File::read calls like 5 times when opening up one file (like it was at 2.8.0).

I know it might be tricky, but we actually need some kind of read Timer that is started, and can only be running one at a time at one fragment. On timeout read async, apply text in uithread and unlock the hlEditor (so at that point it is allowed to save from this editor, and gets editable by users). Beside double read by lifecycle - this also eliminates freezing when opening the activity if the file is big.

**Comment by harshad1** [comment]
Instead of loading the file in the background, we can just show a dialog saying the file has been changed and asking the user if they want to reload. If they do, then we can do a blocking load. This is kind of how vim does it iirc.

I think with 2.8.2 we have 1 load or save at a time. I will investigate removing the modtime.

**Comment by gsantner** [comment]
>Instead of loading the file in the background, we can just show a dialog saying the file has been changed and asking the user if they want to reload. If they do, then we can do a blocking load. This is kind of how vim does it iirc.

no, please not. dialogs are especially tedious on main sceen QuickNote/ToDo

**Comment by gsantner** [comment]
>Instead of loading the file in the background, we can just show a dialog saying the file has been changed and asking the user if they want to reload. If they do, then we can do a blocking load. This is kind of how vim does it iirc.

preferably not. Dialogs are especially bad on main screen QuickNote/ToDo. There could be both changed at same time, so two dialogs popping at same time xD. BUT: Loading indicator / progress overlay in fragment?

> Is this an actual issue? I edit 50-100k character files all the time and I have no issues (but I am on a pixel 3 ... which is a fast phone)

Yes. The bigger the file is, or the worse the phone specs are. I can also notice it when phone is in ultra-battery-saving-mode.

Though, it's not forcefully in scope of PR to load async. I just wanted to share thoughts.

**Pull request closed without merging**